### PR TITLE
Update GetElement and ToScalar to lower IND, LCL_FLD, and LCL_VAR specially

### LINKED ISSUE #86345: Consider decomposing `GetElement(memory, ...)` into equivalent nodes
One motivating example from #85871:

```scala
***** BB02
STMT00005 ( 0x021[E-] ... 0x031 )
               [000025] -----------                         *  JTRUE     void  
               [000024] N--------U-                         \--*  NE        int   
               [000022] -----------                            +--*  CAST      double <- float
               [000053] -----------                            |  \--*  HWINTRINSIC float  float GetElement
               [000020] -----------                            |     +--*  LCL_VAR   simd16<System.Numerics.Vector4>(AX) V00 loc0         
               [000052] -----------                            |     \--*  CNS_INT   int    1
               [000023] -----------                            \--*  CNS_DBL   double 0.0000000000000000
```
We don't get containment for the complex `GetElement` tree. We could if we lowered it as a `LCL_FLD`.

Similar (but not identical) situation exists for `GetElement` sources that are `IND`s:
```
[MethodImpl(MethodImplOptions.NoInlining)]
private static int Problem(Vector128<float> a)
{
    return (int)a.GetElement(1);
}
```

@tannergooding

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves #86345 by updating the lowering of `GetElement`/`ToScalar` to specially handle `GT_IND`, `GT_LCL_FLD`, and `GT_LCL_VAR`.

In particular we see many diffs such as:
```diff
        cmp      edx, 4
        jae      SHORT G_M33861_IG04
-       vmovups  xmm0, xmmword ptr [rcx]
-       vmovaps  xmmword ptr [rsp+20H], xmm0
-       vmovss   xmm0, dword ptr [rsp+4*rdx+20H]
+       vmovss   xmm0, dword ptr [rcx+4*rdx]
```

### CODE REVIEW

**Comment by tannergooding** [comment]
CC. @dotnet/jit-contrib

File: src/coreclr/jit/lowerxarch.cpp [hunk 1]
```diff
@@ -3297,50 +3280,197 @@ void Lowering::LowerHWIntrinsicGetElement(GenTreeHWIntrinsic* node)
...
+
+            if (op2->OperIsConst())
+            {
+                // op2 is a constant, so add it to the existing offset
+
+                BlockRange().Remove(addrMode);
+                BlockRange().Remove(op2);
+
+                int32_t addOffset = (static_cast<uint8_t>(op2->AsIntCon()->IconValue()) % count);
+                addOffset *= static_cast<int32_t>(elemSize);
```
Comment (EgorBo) [inline src/coreclr/jit/lowerxarch.cpp:3325]: is it possible to go out of bounds here?

File: src/coreclr/jit/lowerxarch.cpp [hunk 2]
```diff
@@ -3297,50 +3280,197 @@ void Lowering::LowerHWIntrinsicGetElement(GenTreeHWIntrinsic* node)
...
+            }
+            else if (newIndex == nullptr)
+            {
+                // op2 is not a constant and the addressing mode doesn't
+                // have its own existing index, so use our index and scale
+
+                BlockRange().Remove(addrMode);
+
+                newIndex = op2;
+                newScale = elemSize;
```
Comment (EgorBo) [inline src/coreclr/jit/lowerxarch.cpp:3337]: Is it worth leaving an assert that newScale was 0 (or 1) before?

File: src/coreclr/jit/lowerxarch.cpp [hunk 3]
```diff
@@ -4712,10 +4849,67 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
...
+        // We will specially handle ToScalar when op1 is already in memory
+
+        if (op1->OperIs(GT_IND))
+        {
+            // We want to optimize ToScalar down to an Indir where possible as
+            // this unlocks additional containment opportunities for various nodes
+
+            GenTreeIndir* indir = op1->AsIndir();
+
+            GenTreeIndir* newIndir = comp->gtNewIndir(simdBaseType, indir->Addr(), (indir->gtFlags & GTF_IND_FLAGS));
```
Comment (EgorBo) [inline src/coreclr/jit/lowerxarch.cpp:4863]: can we possibly ignore Scale/Index from the existing `indir` when we create a new one?

File: src/coreclr/jit/lowerxarch.cpp [hunk 4]
```diff
@@ -4712,10 +4849,67 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
...
+
+        if (op1->OperIs(GT_LCL_VAR, GT_LCL_FLD))
+        {
+            uint32_t elemSize = genTypeSize(simdBaseType);
+
+            // We want to optimize ToScalar down to a LclFld where possible as
+            // this unlocks additional containment opportunities for various nodes
+
+            GenTreeLclVarCommon* lclVar  = op1->AsLclVarCommon();
+            uint32_t             lclOffs = lclVar->GetLclOffs() + (0 * elemSize);
```
Comment (EgorBo) [inline src/coreclr/jit/lowerxarch.cpp:4886]: `0 * elemSize` is for readability?

**Review by EgorBo — APPROVED** [review]
LGTM along with a few minor questions

**Review by EgorBo — APPROVED** [review]
LGTM along with a few minor questions

File: src/coreclr/jit/lowerxarch.cpp (hunk 1 above)
Comment (tannergooding) [reply src/coreclr/jit/lowerxarch.cpp:3325]: The constant is restricted to between `0..255` and then `% count` That means its already never going to be more than 64. `Vector512<byte>` is 63, `Vector512<long>` is `7 * 8 == 56`, etc

Combined with the range check we've already emitted for non-constant indices where required means we should always end up with something that is in range.

Except maybe there's still some edge case for the existing offset being `int.MaxValue` though? Maybe `*(basePtr + int.MaxValue).GetElement(5)`....

File: src/coreclr/jit/lowerxarch.cpp (hunk 2 above)
Comment (tannergooding) [reply src/coreclr/jit/lowerxarch.cpp:3337]: Adding an assert would require us to initialize it to zero explicitly. As it is right now this is the definite assignment of the local for this path.

File: src/coreclr/jit/lowerxarch.cpp (hunk 3 above)
Comment (tannergooding) [reply src/coreclr/jit/lowerxarch.cpp:4863]: Not sure what you mean? The scale/index are part of the `Addr` which we copy over from the previous `Indir`

File: src/coreclr/jit/lowerxarch.cpp (hunk 4 above)
Comment (tannergooding) [reply src/coreclr/jit/lowerxarch.cpp:4886]: Yeah, just making it explicitly that we accounted for the index being zero and so no change to the offset is needed.

We do still need to ensure we aren't escaping the bounds of the local with the read, however.

File: src/coreclr/jit/lowerxarch.cpp (hunk 3 above)
Comment (EgorBo) [reply src/coreclr/jit/lowerxarch.cpp:4863]: Ah, nvm, indeed

File: src/coreclr/jit/lowerxarch.cpp (hunk 1 above)
Comment (tannergooding) [reply src/coreclr/jit/lowerxarch.cpp:3325]: and logged https://github.com/dotnet/runtime/issues/86609 to track an implicit truncation that `GenTreeAddrMode` is doing

**Pull request merged**